src/random_bq_to_anndata.py

```diff
@@ -1,46 +1,178 @@
 
-def get_cell_data(project, dataset, client, num_cells):
-    random_ids = get_random_ids(project, dataset, client, num_cells)
-    in_clause = f" matrix.cas_cell_index IN ({','.join(map(str, random_ids))})"
+# Retrieve metadata for the selected (random) cells from the database
+def get_cells(project, dataset, client, random_cell_ids):
+    in_clause = f" cas_cell_index IN ({','.join(map(str, random_cell_ids))})"
+    sql = f"SELECT cas_cell_index, original_cell_id, cell_type FROM `{project}.{dataset}.cas_cell_info` WHERE " + in_clause + " ORDER BY cas_cell_index"
+    query = client.query(sql)
+    ordered_cell_indices = []
+    cell_index_to_cell = {}
+    for row in query:
+        index = row["cas_cell_index"]
+        ordered_cell_indices.append(index)
+        cell_index_to_cell[index] = Cell(row["cas_cell_index"], row["original_cell_id"], row["cell_type"])
+    return ordered_cell_indices, cell_index_to_cell
+
+def get_matrix_data(project, dataset, client, random_cell_ids):
+    in_clause = f" matrix.cas_cell_index IN ({','.join(map(str, random_cell_ids))})"
 
     # at some point, we will probably want create temp table of cell_ids and then JOIN on it
     # instead of an IN clause
-    sql = f"SELECT original_cell_id, cell_type, original_feature_id, feature_name, raw_counts AS count FROM `{project}.{dataset}.cas_cell_info` AS cell, `{project}.{dataset}.cas_feature_info` AS feature, `{project}.{dataset}.cas_raw_count_matrix` AS matrix WHERE matrix.cas_cell_index = cell.cas_cell_index AND matrix.cas_feature_index = feature.cas_feature_index AND" + in_clause
-    print(f"Getting {num_cells} random cells' data from {project}.{dataset}...")
+    # NOTE - really don't need to join to cas_cell_index and cas_feature_index here anymore.
+    sql = f"SELECT matrix.cas_cell_index, original_cell_id, matrix.cas_feature_index, original_feature_id, raw_counts AS count FROM `{project}.{dataset}.cas_cell_info` AS cell, `{project}.{dataset}.cas_feature_info` AS feature, `{project}.{dataset}.cas_raw_count_matrix` AS matrix WHERE matrix.cas_cell_index = cell.cas_cell_index AND matrix.cas_feature_index = feature.cas_feature_index AND" + in_clause + " ORDER BY matrix.cas_cell_index, matrix.cas_feature_index"
     query = client.query(sql)
     return query.result()
 
 
-def random_bq_to_anndata(project, dataset, num_cells):
+def random_bq_to_anndata(project, dataset, num_cells, output_file_prefix):
     client = bigquery.Client(project=project)
-    cell_data = get_cell_data(project, dataset, client, num_cells)
-    adata = ad.AnnData()
-    for row in list(cell_data):
-        # insert VS-507 here
-        pass
+
+    print(f"Getting {num_cells} random cells' data from {project}.{dataset}...")
+    random_cell_ids = get_random_cell_ids(project, dataset, client, num_cells)
+
+    # Read the cell information and store metadata for later
+    ordered_cell_indices, cell_index_to_cell = get_cells(project, dataset, client, random_cell_ids)
```

Comment:
could this just return ordered `Cell`s instead?

Reply:
Yup, good idea.